Topbar5

abenari opened the pull request:
Plugin interface for registering a plugin
including
name, url, description, author, author url,
adding permission
adding roles
manipulating menu items.

example plugin can look like this one:

``` ruby
Foreman::Plugin.register :discovery do
        name 'Discovery plugin'
        author 'Greg Sutcliffe'
        description 'MaaS Discovery Plugin engine for Foreman'
        version VERSION
        requires_foreman '1.3'

        # Add permissions
        security_block :discovery do
          permission :perform_discovery, {:discovers => 
                                [:index, :show, :update, :new, :edit, :destroy, 
                                 :create, :auto_complete_search,
                                 :update_multiple_organization,
                                 :select_multiple_location,
                                 :refresh_facts,
                                 :update_multiple_location,
                                 :submit_multiple_destroy, 
                                 :multiple_destroy]}
        end

        # Add a new role called 'Discovery' if it doesn't exist
        role "Discovery", [:perform_discovery]

        #add menu entry
        menu :top_menu, :discovery, :url_hash => {:controller=> :discovers, :action=>:index},
                                    :caption=> 'Discovered Hosts',
                                    :parent => :hosts_menu,
                                    :after=>:hosts
      end
```

skottler commented:
Tests are failing on 1.8.7.

domcleal commented:
<pre>
/var/lib/workspace/workspace/test_develop_pull_request/database/postgresql/puppet/3.0/ruby/1.8.7/app/services/menu/menu_node.rb:56:in `add_at': Child already added (RuntimeError)
    from /var/lib/workspace/workspace/test_develop_pull_request/database/postgresql/puppet/3.0/ruby/1.8.7/app/services/menu/menu_node.rb:73:in `add'
    from /var/lib/workspace/workspace/test_develop_pull_request/database/postgresql/puppet/3.0/ruby/1.8.7/app/services/menu/menu_manager.rb:73:in `push'
    from /var/lib/workspace/workspace/test_develop_pull_request/database/postgresql/puppet/3.0/ruby/1.8.7/app/services/menu/menu_manager.rb:78:in `item'
    from /var/lib/workspace/workspace/test_develop_pull_request/database/postgresql/puppet/3.0/ruby/1.8.7/app/services/menu/menu_loader.rb:2
    from /var/lib/workspace/workspace/test_develop_pull_request/database/postgresql/puppet/3.0/ruby/1.8.7/app/services/menu/menu_manager.rb:25:in `map'
    from /var/lib/workspace/workspace/test_develop_pull_request/database/postgresql/puppet/3.0/ruby/1.8.7/app/services/menu/menu_loader.rb:1
</pre>

lzap commented on app/services/foreman/plugin.rb:
```diff
@@ -0,0 +1,170 @@
+    def requires_foreman(matcher)
+      current = SETTINGS[:version].gsub('-','.')
+      unless Gem::Dependency.new(nil, matcher).match?(nil, current)
+        raise PluginRequirementError.new("#{id} plugin requires Foreman #{matcher} but current is #{current}")
```
Can we make this just a warning? Because it looks like we are only able to compare fixed versions. We will have lots of plugins what could work just fine with future releases of Foreman and until the version is bumped here, users will be blocked from using them. Just an idea.

lzap commented on app/views/about/index.html.erb:
```diff
@@ -23,65 +20,82 @@
+                <th><%= _("Name") %></th>
+                <th><%= _("Description") %></th>
+                <th><%= _("Author") %></th>
+                <th><%= _("Version") %></th>
```
Instead of rising an error, we could make the version here in red color or something. Looking nice!

ehelms commented:
Moving here since the previous PR was closed. I ask for consideration/discussion of a layout based on - https://gist.github.com/ehelms/7140844 which was generated after some discussions with @domcleal and Kyle Baker.

domcleal commented on app/services/menu/loader.rb:
```diff
@@ -0,0 +1,64 @@
+
+  menu.sub_menu :infrastructure_menu,  :caption => _('Infrastructure') do
+    menu.item :smart_proxies,     :caption => _('Smart proxies')
+    menu.item :compute_resources, :caption => _('Compute resources')
```
Except for smart proxies, these are hidden in unattended: false mode today, probably still should be?

domcleal commented on app/services/foreman/plugin.rb:
```diff
@@ -0,0 +1,179 @@
+module Foreman #:nodoc:
+
+  class PluginNotFound < StandardError; end
+  class PluginRequirementError < StandardError; end
```
These should be Foreman::Exception subclasses

domcleal commented on app/services/foreman/plugin.rb:
```diff
@@ -0,0 +1,179 @@
+    def requires_foreman(matcher)
+      current = SETTINGS[:version].to_s.gsub('-', '.')
+      unless Gem::Dependency.new('', matcher).match?('', current)
+        raise PluginRequirementError.new("#{id} plugin requires Foreman #{matcher} but current is #{current}")
```
When using Foreman::Exception, these errors (and below) can be localised

domcleal commented:
:+1: good improvements I think.  I'll leave others to comment on the menu design, but the plugin registration is excellent.

Some more comments:
1. when logging in as a user with minimal rights, there are menus shown without any entries which should be hidden, e.g. Monitor, Configure and Infrastructure.
2. also unprivileged users see a divider line at the top of the Administer menu which should be removed.
3. plugin information should be exposed over the API, similar to /api/status?
4. plugin information must be collected by scripts/foreman-debug.  Perhaps add a rake task to print registered plugins, and we can execute "foreman-rake plugins:list" or similar from foreman-debug.
5. sorting of menus appears to be deliberate, rather than alphabetical, am I understanding correctly?
6. it throws an error at the moment if you set unattended: false:

<pre>
ActionView::Template::Error (class or module required for rescue clause):
  app/services/menu/item.rb:31:in `rescue in authorized?'
  app/services/menu/item.rb:27:in `authorized?'
  app/services/menu/node.rb:28:in `block in authorized_children'
  app/services/menu/node.rb:28:in `map'
  app/services/menu/node.rb:28:in `authorized_children'
  app/views/home/_topbar.html.erb:19:in `block (2 levels) in _app_views_home__topbar_html_erb___2957712969094668234_57553420'
  app/views/home/_topbar.html.erb:18:in `each'
  app/views/home/_topbar.html.erb:18:in `block in _app_views_home__topbar_html_erb___2957712969094668234_57553420'
  app/views/home/_topbar.html.erb:14:in `_app_views_home__topbar_html_erb___2957712969094668234_57553420'
  app/views/layouts/application.html.erb:32:in `_app_views_layouts_application_html_erb__2112554721630301536_41600600'
  app/controllers/dashboard_controller.rb:6:in `index'
  app/models/concerns/foreman/thread_session.rb:33:in `clear_thread'
</pre>

domcleal commented on app/views/about/index.html.erb:
```diff
@@ -23,65 +20,82 @@
+              <% @plugins.each do |plugin| %>
+                <tr>
+                  <td><%= plugin.url.blank? ? plugin.name : link_to(plugin.name, plugin.url, :rel=>'external')%></td>
+                  <td><%= plugin.description %></td>
```
Perhaps do `<%= _(plugin.description) %>` which will allow plugin authors to translate the descriptions.

domcleal replied:
Same for plugin.name if it's free text (not the gem name), which appears to be the case.

domcleal commented on app/services/menu/loader.rb:
```diff
@@ -0,0 +1,64 @@
+Menu::Manager.map :admin_menu do |menu|
+  menu.item :locations,           :caption => _('Locations') if SETTINGS[:locations_enabled]
```
All uses of `_()` here need to change to `N_()` so they're not translated at this point in the execution.  I think here, we'll be in a Rails initialiser, so the current locale isn't known.  `N_()` will just extract the strings but leave them in English.

When you come to render the menu names and items, that's when you pass the string through `_()` and it will translate.

domcleal commented on app/helpers/home_helper.rb:
```diff
@@ -1,91 +1,31 @@
-      link_to_if_authorized(label, path)
-    end
+  def menu_item_tag item
+    content_tag(:li, link_to(item.caption, item.url_hash), :class => "menu_tab_#{item.url_hash[:controller]} ")
```
`_(item.caption)`

domcleal commented on app/views/home/_submenu.html.erb:
```diff
@@ -0,0 +1,23 @@
+          <%= menu_item_tag item %>
+      <% when Menu::Divider %>
+        <%= content_tag(:li, "", :class => "divider") %>
+        <%= content_tag(:li, item.caption, :class => "nav-header") if item.caption %>
```
`_(item.caption)`

domcleal commented on app/views/home/_submenu.html.erb:
```diff
@@ -0,0 +1,23 @@
+        <%= content_tag(:li, item.caption, :class => "nav-header") if item.caption %>
+      <% when Menu::Toggle %>
+        <%= content_tag :li, :class => "dropdown-submenu" do %>
+          <%= link_to(item.caption, "#", :class => "dropdown-toggle", :"data-toggle" => "dropdown") %>
```
`_(item.caption)`

domcleal commented on app/views/home/_topbar.html.erb:
```diff
@@ -8,26 +8,22 @@
-  </div>
+        <%= render "home/org_switcher" %>
+        <% top_menu_items.each do |menu| %>
+          <%= render "home/submenu", :menu_items => menu.authorized_children, :menu_title => menu.caption %>
```
`_(menu.caption)`

ehelms commented:
On Fri, Nov 1, 2013 at 8:02 AM, Dominic Cleal notifications@github.comwrote:

> [image: :+1:] good improvements I think. I'll leave others to comment on
> the menu design, but the plugin registration is excellent.
> 
> Some more comments:
> 1. when logging in as a user with minimal rights, there are menus
>    shown without any entries which should be hidden, e.g. Monitor, Configure
>    and Infrastructure.
> 2. also unprivileged users see a divider line at the top of the
>    Administer menu which should be removed.
> 3. plugin information should be exposed over the API, similar to
>    /api/status?
> 
> If you mean a way to query the server and get back a list of installed
> plugins, big +1
> 1. plugin information must be collected by scripts/foreman-debug.
>    Perhaps add a rake task to print registered plugins, and we can execute
>    "foreman-rake plugins:list" or similar from foreman-debug.
> 2. sorting of menus appears to be deliberate, rather than
>    alphabetical, is that correct?
> 3. it throws an error at the moment if you set unattended: false:
> 
> ActionView::Template::Error (class or module required for rescue clause):
>   app/services/menu/item.rb:31:in `rescue in authorized?'
>   app/services/menu/item.rb:27:in`authorized?'
>   app/services/menu/node.rb:28:in `block in authorized_children'
>   app/services/menu/node.rb:28:in`map'
>   app/services/menu/node.rb:28:in `authorized_children'
>   app/views/home/_topbar.html.erb:19:in`block (2 levels) in _app_views_home__topbar_html_erb___2957712969094668234_57553420'
>   app/views/home/_topbar.html.erb:18:in `each'
>   app/views/home/_topbar.html.erb:18:in`block in _app_views_home__topbar_html_erb___2957712969094668234_57553420'
>   app/views/home/_topbar.html.erb:14:in `_app_views_home__topbar_html_erb___2957712969094668234_57553420'
>   app/views/layouts/application.html.erb:32:in`_app_views_layouts_application_html_erb__2112554721630301536_41600600'
>   app/controllers/dashboard_controller.rb:6:in `index'
>   app/models/concerns/foreman/thread_session.rb:33:in`clear_thread'
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/theforeman/foreman/pull/987#issuecomment-27561189
> .

## 

Go Pack and War Eagle,
Eric D. Helms

Ph.D. Student - North Carolina State University

domcleal commented:
@ehelms yes, I do - we should have UI and API feature parity, ideally from the beginning.

domcleal commented on app/services/foreman/plugin.rb:
```diff
@@ -0,0 +1,179 @@
+    def requires_foreman(matcher)
+      current = SETTINGS[:version].to_s.gsub('-', '.')
+      unless Gem::Dependency.new('', matcher).match?('', current)
+        raise PluginRequirementError.new(_("%{id} plugin requires Foreman %{matcher} but current is %{current}" % {:id=>id, :matcher => matcher, :current=>current}))
```
These should be N_() again.

domcleal commented on app/services/foreman/plugin.rb:
```diff
@@ -0,0 +1,179 @@
+      plugin = Plugin.find(plugin_name)
+      raise PluginRequirementError.new(_("%{id} plugin requires the %{plugin_name} plugin, not found") % {:id =>id, :plugin_name=>plugin_name}) unless plugin
+      unless Gem::Dependency.new('', matcher).match?('', plugin.version)
+        raise PluginRequirementError.new(_("%{id} plugin requires the %{plugin_name} plugin %{matcher} but current is %{plugin_version}" % {:id=>id, :plugin_name=>plugin_name,:matcher=> matcher,:plubin_version=>plugin.version}))
```
s/plubin/plugin/

domcleal commented on app/views/api/v2/plugins/index.json.rabl:
```diff
@@ -0,0 +1,4 @@
+collection @plugins
+
+attributes :name, :author, :description, :url, :version
```
Missing the "id"?  (ID is gem name and "name" is a friendly name, right?)

domcleal commented:
Getting a few test failures appear, could you check?  I suspect you'll need to add the plugins API to access_permissions too (the test failures are probably masking those).  A functional test for the API would be appreciated so we exercise it a little.

An addition to scripts/foreman-debug is the only thing I see missing now, it's important that we capture this data for debugging.

abenari commented:
[test]

abenari commented:
[test]

lzap commented:
In regard of the foreman-debug, go ahead and merge, I will do another PR with change for that.

domcleal commented:
Sorry this is held up on me at the moment with the versioning, will get back to it.

I was just talking to @omaciel who indicated that his team would really appreciate consistent IDs added to menu + submenu items so they can automate testing of Foreman.  Could we do this automatically based off of the menu metadata?

GregSutcliffe commented:
consistent ids would allow us to rewrite the integration tests in such a way that Discovery doesn't break them too (by looking for one match for "Hosts" and finding 2)

abenari committed:
fixes #3510 plugin interface for registring a plugin
including
name url, description, author, author url,
adding permission
adding roles
manipulating menu items.

domcleal commented:
Thanks @abenari, and sorry for the delay while I got the version bits merged.  Merged as 611fd58 for Foreman 1.4.0.
